File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const hash = {};
    
    genres.map((genre, i) => [genre, plays[i]])
    // [
    //   [ 'classic', 500 ],
    //   [ 'pop', 600 ],
    //   [ 'classic', 150 ],
    //   [ 'classic', 800 ],
    //   [ 'pop', 2500 ]
    // ]
          .forEach(([genre, play], index) => {
            const data = hash[genre] || { total: 0, songs: [] }; // hash에 저장되어있으면 불러오고, 없으면 {total: 0, songs: []}로 값 추가
            
            hash[genre] = {
                total: data.total + play, // total에 재생수 더하기
                songs: [...data.songs, {play, index}] // songs 배열에 현재 곡을 {play: 500, index: 0}의 형태로 저장하기
                        .sort((a, b) => b.play - a.play) // 재생수 많은 순으로 내림차순 정렬
                        .slice(0, 2) // 베스트 2개만 뽑기
            }
    })
    // console.log(hash)
    // {
    //   classic: { total: 1450, songs: [  { play: 800, index: 3 }, { play: 500, index: 0 } ] },
    //   pop: { total: 3100, songs: [ { play: 2500, index: 4 }, { play: 600, index: 1 } ] }
    // }
    
    // console.log(Object.entries(hash))
    // [
    //   [ 'classic', { total: 1450, songs: [Array] } ],
    //   [ 'pop', { total: 3100, songs: [Array] } ]
    // ]
    
    return Object.entries(hash).sort((a, b) => b[1].total - a[1].total)
                               .flatMap((el) => el[1].songs) // [{ play: 2500, index: 4 }, { play: 600, index: 1 }, { play: 800, index: 3 }, { play: 500, index: 0 }]
                               .map((song) => song.index);
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

# 장르 별로 가장 많이 재생된 노래 2개씩 리턴
# { classic: {
#       total_plays: 1450,
#       songs: [ {number: 0, plays: 500 }, ... ]
#   }, 
#   ... 
# }
    
def solution(genres, plays):
    answer = []
    genre_map = defaultdict(lambda: {'total_plays': 0, 'songs': []})
    
    for i, (genre, play) in enumerate(zip(genres, plays)):
        genre_map[genre]['total_plays'] += play
        genre_map[genre]['songs'].append({'number': i, 'play': play})
    
    sorted_genres = sorted(genre_map.values(), key=lambda x: -x['total_plays'])
    
    for genre in sorted_genres:
        top_2_songs = sorted(genre['songs'], key=lambda x: (-x['play'], x['number']))[:2]
        answer.extend([song['number'] for song in top_2_songs])
    
    return answer